### ISSUE TITLE
[Meta Issue]: OpenAPI Support 

### ISSUE THREAD

**Issue by sennyeya** [backstage/backstage#17482]
### 🔖 Feature description

This is a collection of issues stemming from #2566. The goal of this ticket is to break apart the different types of work from that original ticket into separate camps and define a clearer roadmap for OpenAPI support in the catalog.

## What do you mean by OpenAPI support in the catalog?
This can mean a multitude of things, which is partially why we want to replace the old ticket. Here it will mean
1. OpenAPI specs for plugins.
2. An OpenAPI spec-first developer experience, using the spec to drive server-side code writing.
3. A Backstage-only client generated using the spec.
4. Other language clients like Java, Python for interfacing with Backstage plugins.

## Plugin Specs
- [x] catalog-backend (#15667, #16238)

## Spec-first Development
- [x] Typed express router methods (#15667)
- [ ] Request, params and response validation
- [ ] Auto-generated server stubs

## Internal Client Autogeneration
- [x] Custom generator from `openapi-generator` (#17470)
- [ ] Migrate existing clients to use autogenerated versions
  - [ ] catalog-backend
  - [ ] others

## Additional Client Autogeneration
- [ ] Support for additional languages
  - [ ] Java
  - [ ] Python
  - [ ] Typescript

### Open Questions
- Should we have a single monolithic client that includes all plugins and specs or should each plugin have its own autogenerated client?

## Others
- [ ] Serving these specs in Backstage for _meta_ discovery.
- [ ] `.well-known/openapi.yaml` endpoint with a runtime generated API spec

### 🎤 Context

n/a

### 👀 Have you spent some time to check if this feature request has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/backstage/backstage/blob/master/CODE_OF_CONDUCT.md)

### Are you willing to submit PR?

None

**Comment by sennyeya** [comment]
Also, found this library for jest that adds spec validation into the responses, https://github.com/openapi-library/OpenAPIValidators/tree/master/packages/jest-openapi#readme. Would be useful for verifying response values in test cases as a precursor to turning on `validateResponses` in the middleware.

**Comment by dannysheridan** [comment]
@sennyeya as you think about the client autogeneration, you ought to check out [Fern](www.buildwithfern.com)

**Comment by Rugvip** [comment]
@dannysheridan apart from discussion on the technical solution, this part of the experience is a blocker for us to adopt Fern as a default tool in the Backstage OSS project:

<img width="290" alt="image" src="https://github.com/backstage/backstage/assets/4984472/47cbc863-c720-41fd-a195-e91104006afa">

**Comment by dsinghvi** [comment]
> @dannysheridan apart from discussion on the technical solution, this part of the experience is a blocker for us to adopt Fern as a default tool in the Backstage OSS project:
> 
> <img alt="image" width="290" src="https://user-images.githubusercontent.com/4984472/244377781-47cbc863-c720-41fd-a195-e91104006afa.png">

@Rugvip we can add a `--skip-login` option on init/generate. Would that unblock adopting fern? (So far enforcing login has been great for analytics/support but I see why that's extremely limiting in an OSS project).

**Comment by Rugvip** [comment]
@dsinghvi If login is not required and there are no usage limitations for the core use-cases we're likely at a point where we wouldn't dismiss it. We'd still ofc need to evaluate whether it's a desirable solution and what level of lock-in there is.

**Comment by Stranger6667** [comment]
@sennyeya if you need anything from the Schemathesis side, feel free to tag me and I’ll be happy to provide support on integrating it :)

**Comment by sennyeya** [comment]
@Stranger6667 That was quick! Will definitely keep you in the loop as I start exploring. Thanks for the great tool!

**Comment by vinzscam** [comment]
hey have we considered finding a way to not run `openapi generate-client` script when developing?

I felt the pain of writing yaml, then run a script to generate a `.ts` file which is then consumed by our typescript bundler.
I found the process really slow and is yet another thing that contributors need to learn before contributing to Backstage.

I would love to just write the openapi definition in typescript instead.

**Comment by sennyeya** [comment]
@vinzscam Reached out via DM, happy to help but I'd like a little more info on your use case.

**Comment by vinzscam** [comment]
> @vinzscam Reached out via DM, happy to help but I'd like a little more info on your use case.

nothing fancy I am just developing a plugin which integrates with OpenAPI and found the process a bit complicated 😅

I'm thinking that writing the definition directly in typescript (for example by using some [zod](https://www.npmjs.com/package/zod) to openapi library) might simplify the development process.

**Comment by ScottGuymer** [comment]
I have been trying to use the spec first process to build some plugins.

It really is a great way forward for building plugins, especially when combined with the catalog plugin to expose the API specs in the catalog.

However one issue I have run into seems to be that the client generated is a little too simple.

It doesn't seem to take account of errors or different return codes for an endpoint and just seems to deal with the success (2xx).

It seems to generate the models for other responses (400, 401 etc) but never actually uses them in the default client.

Am I missing something? Or is this just the way it works right now? FWIW the backend seems to do this correctly.

**Comment by aramissennyeydd** [comment]
@ScottGuymer 👋 Awesome to hear you're using this!

> However one issue I have run into seems to be that the client generated is a little too simple.

AFAIK we can't easily type failure cases without causing weird issues for the returned interface - open to improvements here though! The client generation is trying to balance a simple `res.json` API and still having errors available. A lot of this was designed to fit into (and eventually replaced) the existing `CatalogClient` implementation, https://github.com/backstage/backstage/blob/master/packages/catalog-client/src/CatalogClient.ts#L368-L381.

For the backend, it's a bit easier as we can discriminate based on status code _first_ to enforce specific types per the spec.

**Comment by ScottGuymer** [comment]
Fwiw I also use hey-api for generating clients and that does take care of various return types specified.

Not looked into the mechanics of how that works under the hood enough to comment.

Also not sure if the octokit client from GH is generated too. Maybe not from openapi spec though

**Comment by ScottGuymer** [comment]
Another issue I'm running into is the combination of generating the openapi-client and running api-reports (like we do in a plugin).

The documentation for the query string types is not being correctly passed through from the openapi spec into the genrated api client.

This then causes multiple `// src/schema/openapi/generated/apis/Api.client.d.ts:82:1 - (ae-undocumented) Missing documentation for "DefaultApiClient".` to be shown in the api-report if any of the generated code is shown in the external APi surface.

You can kind of work around this by haveing a wrapper for the default API client and only expose the models (assuming you have documented them correctly)

Im not sure if this is expected behaviour or not?

**Comment by aramissennyeydd** [comment]
@ScottGuymer Can you share an example plugin where this is happening? I'd expect this block to allow that check to pass - https://github.com/backstage/backstage/blob/2a8b0af434910caaff47baf2f4183b1b22bf470d/packages/repo-tools/templates/typescript-backstage-client/apis/api-all.mustache#L69-L72

**Comment by ScottGuymer** [comment]
The example I have is an internal plugin im afraid.

But we are not doing anything special in the setup here afaik..

I see the exact tsdoc as you have highlighted here on the `DefaultApiClient`.

It also complains about any parameter objects used in the default API that represent query strings in the api spec. They have the same `@public` annotation, but none of the documentation is passed through to the members from the spec.

example

<img width="345" height="239" alt="Image" src="https://github.com/user-attachments/assets/4fb1bad4-a6b0-4e6f-a37c-ce41cf5aa3fb" />
